c2rust-analyze/src/known_fn.rs

```diff
@@ -0,0 +1,362 @@
+use crate::context::PermissionSet;
+
+#[cfg(test)]
+macro_rules! const_slice {
+    ($ty:ty, []) => {{
+        &[]
+    }};
+    ($ty:ty, $array:expr) => {{
+        const ARRAY: [$ty; $array.len()] = $array;
+        &ARRAY
+    }};
+}
+
+#[cfg(test)]
+macro_rules! perms_annotation {
+    ([$($($perm:ident)|*),*]) => {{
+        [$(PermissionSet::union_all([$(PermissionSet::$perm,)*]),)*]
+    }};
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct KnownFnTy {
+    name: &'static str,
+    ty: &'static str,
+    perms: &'static [PermissionSet],
+}
+
+impl KnownFnTy {
+    /// Check that we annotated the right number of [`PermissionSet`]s
+    /// that corresponds to the number of raw pointers in [`Self::ty`].
+    #[cfg(test)]
+    const fn checked(self) -> Self {
+        let ty = self.ty.as_bytes();
+        let mut num_ptrs = 0;
+        let mut i = 0;
+        while i < ty.len() {
+            if ty[i] == b'*' {
+                num_ptrs += 1;
+            }
+            i += 1;
+        }
```

Comment:
Why not use a for loop here?

Reply:
Same as https://github.com/immunant/c2rust/pull/978#discussion_r1253750006.